logic/tool/TypeResolver.java

```diff
@@ -397,17 +410,45 @@ private void addNeighbours(TypeVariable from, TypeVariable to) {
-                else if (constraint.isRelates()) addNeighbours(from, constraint.asRelates().role());
-                else if (constraint.isIs()) addNeighbours(from, constraint.asIs().variable());
-                // TODO: There are other constraints in a TypeVariable. Are we intentionally ignoring them?
-                //       Why do we not throw GraknException.of(ILLEGAL_STATE); ?
+        public TypeVariable cloneVariable(TypeVariable copyFrom) {
+            TypeVariable newVar = resolvers.register(copyFrom);
+
+            for (TypeConstraint constraint : copyFrom.constraints()) {
+                if (constraint.isLabel()) {
+                    newVar.label(constraint.asLabel().properLabel());
+                } else if (constraint.isValueType()) {
+                    newVar.valueType(constraint.asValueType().valueType());
+                } else if (constraint.isRegex()) {
+                    newVar.regex(constraint.asRegex().regex());
+                } else if (constraint.isAbstract()) {
+                    newVar.setAbstract();
+                } else if (constraint.isSub()) {
+                    newVar.sub(convert(constraint.asSub().type()), constraint.asSub().isExplicit());
+                    addNeighbours(newVar, constraint.asSub().type());
+                } else if (constraint.isOwns()) {
+                    newVar.owns(convert(constraint.asOwns().attribute()),
+                                constraint.asOwns().overridden().flatMap(var -> Optional.of(convert(var))).orElse(null),
+                                constraint.asOwns().isKey());
+                    addNeighbours(newVar, constraint.asOwns().attribute());
+                    constraint.asOwns().overridden().ifPresent(typeVariable -> addNeighbours(newVar, typeVariable));
+                } else if (constraint.isPlays()) {
+                    newVar.plays(constraint.asPlays().relation().flatMap(var -> Optional.of(convert(var))).orElse(null),
+                                 convert(constraint.asPlays().role()),
+                                 constraint.asPlays().overridden().flatMap(var -> Optional.of(convert(var))).orElse(null));
+                    addNeighbours(newVar, constraint.asPlays().role());
+                    constraint.asPlays().overridden().ifPresent(typeVariable -> addNeighbours(newVar, typeVariable));
+                } else if (constraint.isRelates()) {
+                    newVar.relates(convert(constraint.asRelates().role()),
+                                   constraint.asRelates().overridden().flatMap(var -> Optional.of(convert(var))).orElse(null));
+                    addNeighbours(newVar, constraint.asRelates().role());
+                    constraint.asRelates().overridden().ifPresent(typeVariable -> addNeighbours(newVar, typeVariable));
+                } else if (constraint.isIs()) {
+                    newVar.is(convert(constraint.asIs().variable()));
+                } else throw GraknException.of(ILLEGAL_STATE);
```

Comment:
I think it would be neater if we extract each if-else block into its own method?

Reply:
Agreed, I'm going to do it with the Traversal re-write. This is also make the convert logic look similar for both `ThingVariable` and `TypeVariable`, which is a plus.

Reply:
Okay, sounds good!